src/interface/swarm_boundaries.hpp

```diff
@@ -0,0 +1,139 @@
+//========================================================================================
+// (C) (or copyright) 2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SWARM_BOUNDARIES_HPP_
+#define INTERFACE_SWARM_BOUNDARIES_HPP_
+
+#include "swarm_device_context.hpp"
+
+namespace parthenon {
+
+class ParticleBound {
+ public:
+  KOKKOS_INLINE_FUNCTION virtual void Apply(const int n, double &x, double &y, double &z,
+                                            // const SwarmDeviceContext &context) = 0;
```

Comment:
I thought that would work but when I do that, meaning
```c++
 KOKKOS_INLINE_FUNCTION virtual void Apply(const int n, double &x, double &y, double &z,
                                             const SwarmDeviceContext &context) = 0
 ```
and remove the `override`s from the derived classes the compiler thinks the derived classes are abstract for some reason:
```
/home/brryan/github/parthenon/src/kokkos_abstraction.hpp:769:67: error: invalid new-expression of abstract class type 'parthenon::ParticleBoundIX2Outflow'
  769 |       Kokkos::RangePolicy<ES>(0, 1), KOKKOS_LAMBDA(const int i) { new (p) T(); });
      |                                                                   ^~~~~~~~~~~
In file included from /home/brryan/github/parthenon/src/interface/swarm.hpp:37,
                 from /home/brryan/github/parthenon/src/interface/state_descriptor.hpp:29,
                 from /home/brryan/github/parthenon/src/application_input.hpp:24,
                 from /home/brryan/github/parthenon/src/mesh/mesh.hpp:33,
                 from /home/brryan/github/parthenon/src/interface/swarm.cpp:19:
/home/brryan/github/parthenon/src/interface/swarm_boundaries.hpp:79:7: note:   because the following virtual functions are pure within 'parthenon::ParticleBoundIX2Outflow':
   79 | class ParticleBoundIX2Outflow : public ParticleBound {
      |       ^~~~~~~~~~~~~~~~~~~~~~~
/home/brryan/github/parthenon/src/interface/swarm_boundaries.hpp:22:39: note: 	'virtual void parthenon::ParticleBound::Apply(int, double&, double&, double&, const parthenon::SwarmDeviceContext&)'
   22 |   KOKKOS_INLINE_FUNCTION virtual void Apply(const int n, double &x, double &y, double &z,
      |
```
Do you recognize that?

Reply:
I think you still need the `override` but you can let the base method be pure.